### DISCUSSION TITLE
Annotated deps - big performance hit

### DISCUSSION THREAD

**Discussion by eddy2by4** [fastapi/fastapi#11654] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated
from fastapi import FastAPI, Depends

app = FastAPI()


def my_dep_fn():
    return 2


@app.get("/")
async def root():
    return 1


@app.get("/slow")
async def slow(my_dep=Annotated[int, Depends(my_dep_fn)]):
    return 1
```

### Description

I've noticed that using Annotated dependancies like in the docs, the performance hit is huge (around a 50% performance hit).
I've ran the snipped of code above on my M3 Pro Macbook Pro and the differences are noticeable:

```
wrk -t12 -c300 -d15s http://localhost:8000/
Running 15s test @ http://localhost:8000/
  12 threads and 300 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    13.51ms   28.07ms 648.17ms   98.77%
    Req/Sec     2.26k   198.33     3.34k    78.61%
  406104 requests in 15.04s, 48.41MB read
  Socket errors: connect 0, read 465, write 0, timeout 0
Requests/sec:  27008.01
Transfer/sec:      3.22MB
```

vs

```
wrk -t12 -c300 -d15s http://localhost:8000/slow
Running 15s test @ http://localhost:8000/slow
  12 threads and 300 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    32.03ms   80.36ms   1.23s    97.62%
    Req/Sec     1.15k   117.65     1.74k    71.67%
  206276 requests in 15.07s, 24.59MB read
  Socket errors: connect 0, read 771, write 0, timeout 0
Requests/sec:  13684.37
Transfer/sec:      1.63MB
```

As you can see I just added the annotated dependancy to the function, I am not doing anything with the function and the server is responding in half the speed.

Is this performance hit normal and expected ?
 

### Operating System

macOS

### FastAPI Version

0.111.0

### Pydantic Version

-

### Python Version

3.12

**Comment by JavierSanchezCastro** [comment]
I didn't try your code, but there is a typo in

```python3
my_dep=Annotated[int, Depends(my_dep_fn)]
```

Instead of
```python3
my_dep: Annotated[int, Depends(my_dep_fn)]
```

**Reply by eddy2by4** [reply]
You are right I had a typo.

~~Fixing the typo it went from : 13k -> 23k~~
Edit: this is wrong, it was not the typo (maybe that also contributed), but fixing the typo had the same performance.
I had to make the function `async` to get to 23k req/s.

Without it its around 27k so I guess a bit of overhead there, but I guess acceptable.

**Reply by Kludex** [reply]
If you make the dependency async, what happens?

**Reply by eddy2by4** [reply]
Actually, I was wrong. Sorry. The 23k speed was with the function async (which was different than the initial code submited). With async its still 13k.

So in conclusion:

```
def my_dep_fn():  # performance changes from 27k req/s to 13k req/s
    return 2

async def my_dep_fn():  # performance changes from 27k req/s to 23 req/s
    return 2

@app.get("/slow")
async def slow(my_dep: Annotated[int, Depends(my_dep_fn)]):
    return 1
```


So its still a performance hit, should it happen @Kludex  ?

The major takeaway is that all dependancies should be `async` to maintain close to baseline performance, even if you don't do any async operations in them.
I don't actually know if this is mentioned in the docs, but it might be useful for people to know, since the performance hit is quite significant. Or someone check the code and see if its possible to make it have the same performance with both function types.

**Reply by Kludex** [reply, marked as answer]
It's normal. It's mentioned in the docs that we use a thread pool in non-async functions. That's the penalty on the performance happening here.

**Reply by eddy2by4** [reply]
So for best performance, I should have all my functions in my app `async` ? Even when not doing async work? Or I need to make sure all deps are `async` ?

**Reply by eddy2by4** [reply]
Actually I think I answered it myself (for people reading about this: https://fastapi.tiangolo.com/async/ )

Takeaway: **Path operation functions and dependencies** should always be `async def` ! Normal utility functions can be `def`.


`Path operation functions[¶](https://fastapi.tiangolo.com/async/#path-operation-functions)
When you declare a path operation function with normal def instead of async def, it is run in an external threadpool that is then awaited, instead of being called directly (as it would block the server).

If you are coming from another async framework that does not work in the way described above and you are used to defining trivial compute-only path operation functions with plain def for a tiny performance gain (about 100 nanoseconds), please note that in FastAPI the effect would be quite opposite. In these cases, it's better to use async def unless your path operation functions use code that performs blocking I/O.

Still, in both situations, chances are that FastAPI will [still be faster](https://fastapi.tiangolo.com/#performance) than (or at least comparable to) your previous framework.

Dependencies[¶](https://fastapi.tiangolo.com/async/#dependencies)
The same applies for [dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/). If a dependency is a standard def function instead of async def, it is run in the external threadpool.

Sub-dependencies[¶](https://fastapi.tiangolo.com/async/#sub-dependencies)
You can have multiple dependencies and [sub-dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/) requiring each other (as parameters of the function definitions), some of them might be created with async def and some with normal def. It would still work, and the ones created with normal def would be called on an external thread (from the threadpool) instead of being "awaited".

Other utility functions[¶](https://fastapi.tiangolo.com/async/#other-utility-functions)
Any other utility function that you call directly can be created with normal def or async def and FastAPI won't affect the way you call it.

This is in contrast to the functions that FastAPI calls for you: path operation functions and dependencies.

If your utility function is a normal function with def, it will be called directly (as you write it in your code), not in a threadpool, if the function is created with async def then you should await for that function when you call it in your code.`

**Reply by eddy2by4** [reply]
As an update:

When I ran my actual project that I am developing with Gunicorn and Uvicorn workers. I literally had 0 overhead for using dependancies. The same fast req/s with or without dependancies.

So make no mistake, in a production environment your app will not suffer from any slowdown or performance hit using `async def` Dependancies !

DB:
Without deps: 18000 req/s (round up)
With deps: 18000 req/s (round up)

Redis:
Without deps: 126000 req/s (round up)
With deps: 126400 req/s (round up)

**Comment by eddy2by4** [comment]
@Kludex while we are discussing about Dependency performance, I actually found some documentation that might affect the performance of our users.

https://fastapi.tiangolo.com/advanced/settings/#__tabbed_4_1

Here as you can see the settings example, we have a sync dep:

```
@lru_cache
def get_settings():
    return Settings()


@app.get("/info")
async def info(settings: Annotated[Settings, Depends(get_settings)]):
    return {
        "app_name": settings.app_name,
        "admin_email": settings.admin_email,
        "items_per_user": settings.items_per_user,
    }
```

I followed this example in my app, and surely the routes using this `Annotated[Settings, Depends(get_settings)]` are all affected.

56k req/s vs of 126k req/s (the number I expect)

I understand this is because the `get_settings` function is sync and it runs in a different threadpool. I tried making it async but I got errors relating to "coroutines".

In the end I just used imported the and called the get_settings at the top of all files I need settings for:

```
from core.utils import get_settings
settings = get_settings()

@get("/test")
async def myfn....
```

And now all the endpoints have response times of 126k req/s again.

So this pattern example from the documentation is limiting users performance by more than half, maybe without them knowing.
Can we update the documentation so this does not happen at least for future users ? Also maybe you know a better approach (fastapi way) to solve this problem ?

It surely isn't a good reason to limit app performance for everyone reading variables from env.
Thoughts ?

Thanks!

**Reply by alexandrubese** [reply]
Any thoughts on this ? I am also interested

**Reply by eddy2by4** [reply]
Any update on this ?

**Reply by alexandrubese** [reply]
@Kludex @YuriiMotov @JavierSanchezCastro
